File: tictactoe.r
# Define the initial game board
board <- matrix(rep(" ", 9), ncol = 3)

# Define a function to print the game board
print_board <- function(board) {
  cat(" ", "1", "2", "3", "\n")
  cat("1", board[1,], "\n")
  cat("2", board[2,], "\n")
  cat("3", board[3,], "\n")
}

# Define a function to check if the game is over
game_over <- function(board) {
  # Check for horizontal wins
  for (i in 1:3) {
    if (board[i,1] != " " && all(board[i,] == board[i,1])) {
      return(TRUE)
    }
  }
  
  # Check for vertical wins
  for (j in 1:3) {
    if (board[1,j] != " " && all(board[,j] == board[1,j])) {
      return(TRUE)
    }
  }
  
  # Check for diagonal wins
  if (board[1,1] != " " && all(diag(board) == board[1,1])) {
    return(TRUE)
  }
  if (board[1,3] != " " && all(diag(board[,3:1]) == board[1,3])) {
    return(TRUE)
  }
  
  # Check for a tie
  if (!any(board == " ")) {
    return(TRUE)
  }
  
  # If we get here, the game is not over
  return(FALSE)
}

# Define a function to get the user's move
get_move <- function(player) {
  repeat {
    move <- readline(paste0(player, "'s turn (row,column): "))
    move <- as.numeric(strsplit(move, ",")[[1]])
    if (length(move) == 2 && all(move %in% 1:3) && board[move[1],move[2]] == " ") {
      return(move)
    }
    cat("Invalid move. Please try again.\n")
  }
}

# Define the main game loop
print_board(board)
while (!game_over(board)) {
  # Player X's turn
  move <- get_move("X")
  board[move[1],move[2]] <- "X"
  print_board(board)
  if (game_over(board)) {
    break
  }
  
  # Player O's turn
  move <- get_move("O")
  board[move[1],move[2]] <- "O"
  print_board(board)
}

# Print the final game board and outcome
print_board(board)
if (any(diag(board) == "X") || all(diag(board) == "O")) {
  cat("Player ", board[1,1], " wins!\n")
} else {
  cat("Tie game.\n")
}

File: tictactoe.rb
class TicTacToe
  def initialize
    @board = Array.new(3) { Array.new(3, " ") }
    @current_player = "X"
  end

  def play
    loop do
      display_board
      print_turn_prompt
      row, col = get_move
      make_move(row, col)
      break if game_over?
      switch_players
    end
    display_board
    print_result
  end

  private

  def display_board
    puts "   |   |   "
    puts " #{@board[0][0]} | #{@board[0][1]} | #{@board[0][2]} "
    puts "___|___|___"
    puts "   |   |   "
    puts " #{@board[1][0]} | #{@board[1][1]} | #{@board[1][2]} "
    puts "___|___|___"
    puts "   |   |   "
    puts " #{@board[2][0]} | #{@board[2][1]} | #{@board[2][2]} "
    puts "   |   |   "
  end

  def print_turn_prompt
    puts "#{@current_player}'s turn. Enter row and column (e.g. 1 2):"
  end

  def get_move
    loop do
      input = gets.chomp.split.map(&:to_i)
      row, col = input[0], input[1]
      if valid_move?(row, col)
        return row, col
      else
        puts "Invalid move. Please try again."
      end
    end
  end

  def valid_move?(row, col)
    @board[row - 1][col - 1] == " "
  end

  def make_move(row, col)
    @board[row - 1][col - 1] = @current_player
  end

  def switch_players
    @current_player = @current_player == "X" ? "O" : "X"
  end

  def game_over?
    win? || draw?
  end

  def win?
    rows_win? || cols_win? || diagonals_win?
  end

  def rows_win?
    @board.any? { |row| row.uniq.size == 1 && row[0] != " " }
  end

  def cols_win?
    @board.transpose.any? { |col| col.uniq.size == 1 && col[0] != " " }
  end

  def diagonals_win?
    diagonals = [[@board[0][0], @board[1][1], @board[2][2]],
                 [@board[0][2], @board[1][1], @board[2][0]]]
    diagonals.any? { |diag| diag.uniq.size == 1 && diag[0] != " " }
  end

  def draw?
    @board.flatten.none? { |cell| cell == " " }
  end

  def print_result
    if win?
      puts "#{@current_player} wins!"
    else
      puts "It's a draw."
    end
  end
end

game = TicTacToe.new
game.play
